### PR TITLE
EHPR-74 |  Application Deployment on AWS

### PR DESCRIPTION
**Pull request by fwpushan**
1. Basic Application Deployment on AWS
2. Terraform infra code
3. Building cmd

### CODE REVIEW

File: apps/api/src/lambda.ts [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  APIGatewayProxyResult,
+  Context,
+  Callback,
+  Handler,
+} from 'aws-lambda';
+import { AppModule } from './app.module';
+import { ExpressAdapter } from '@nestjs/platform-express';
+import { API_PREFIX } from './config';
+
+let cachedServer: Handler;
```
Comment (arranfw) [inline apps/api/src/lambda.ts:15]: Cool

File: tsconfig.json [hunk 2]
```diff
@@ -1,15 +1,15 @@
 {
-  "extends": "@tsconfig/node14/tsconfig.json",
   "compilerOptions": {
-    "lib": ["es2020"],
+    "lib": ["esnext"],
```
Comment (arranfw) [inline tsconfig.json:3]: Did these end up being required for deployment to work?

**Review by arranfw — COMMENTED** [review]
Nothing jumps out immediately, let's go over this as a team tomorrow

**Comment by arranfw** [comment]
Seems like we have 3 vars for environments

- RUNTIME_ENV
- ENV_NAME
- NODE_ENV

Do we need all three? Can we document what the variables are and what values map to?

File: tsconfig.json (hunk 2 above)
Comment (fwpushan) [reply tsconfig.json:3]: Nope.....the solution was crazy.

**Comment by fwpushan** [comment]
> Seems like we have 3 vars for environments
> 
> * RUNTIME_ENV
> * ENV_NAME
> * NODE_ENV
> 
> Do we need all three? Can we document what the variables are and what values map to?

@arranfw RUNTIME_ENV is useful to skip env injection. Others are optional right now.

File: terraform/variable.tf [hunk 3]
```diff
@@ -0,0 +1,14 @@
+variable "project_code" {}
+variable "api_artifact" {}
+variable "app_sources" {}
+variable "target_env" {}
+variable "domain" {}
+variable "app_sources_bucket" {}
+
+variable "function_memory_mb" {
+  default = "2048"
```
Comment (dbayly-freshworks) [inline terraform/variable.tf:9]: Nit, but this seems a bit small

File: terraform/main.tf [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+      Environment = var.target_env
+      Project     = var.project_code
+      Author      = "FreshWorks Studio"
+    }
+  }
+}
+
+provider "aws" {
+  alias  = "us-east-1"
+  region = "us-east-1"
```
Comment (dbayly-freshworks) [inline terraform/main.tf:28]: Shouldn't this be ca-central-1?

File: terraform/api.tf [hunk 5]
```diff
@@ -0,0 +1,77 @@
+resource "aws_lambda_function" "api" {
+  description      = "API for ${local.namespace}"
+  function_name    = local.api_name
+  role             = aws_iam_role.lambda.arn
+  runtime          = "nodejs14.x"
```
Comment (dbayly-freshworks) [inline terraform/api.tf:5]: I know currently we're running a lot of node 14, but it might be worth using a newer version

File: Makefile [hunk 6]
```diff
@@ -4,15 +4,63 @@
...
+define TFVARS_DATA
+target_env = "$(ENV_NAME)"
+project_code = "$(PROJECT)"
+api_artifact = "build/api.zip"
+app_sources = "build/app"
+app_sources_bucket = "$(APP_SRC_BUCKET)"
+domain = ""
+endef
+export TFVARS_DATA
+CLOUDFRONT_ID = EH5H3TN595NCP
```
Comment (dbayly-freshworks) [inline Makefile:32]: Is this something we should put into environment variables?

**Review by dbayly-freshworks — APPROVED** [review]
I have a ton of small questions, but overall this looks fine.

File: terraform/main.tf (hunk 4 above)
Comment (vyasworks) [reply terraform/main.tf:28]: we need two providers - for deploying edge lambda - Cloudfront functions

Comment (vyasworks) [reply terraform/main.tf:28]: one for ca-central - as above, this one for edge lambda

**Review by vyasworks — APPROVED** [review]
LGTM, as we are following up on existing tf work we have done.

File: tsconfig.json (hunk 2 above)
Comment (fwpushan) [reply tsconfig.json]: Reverted

File: Makefile (hunk 6 above)
Comment (fwpushan) [reply Makefile:32]: Yes, I'm changing

File: terraform/api.tf (hunk 5 above)
Comment (arranfw) [reply terraform/api.tf:5]: I think 14 is the latest supported by lambda

File: tsconfig.json [hunk 7]
```diff
@@ -9,7 +8,8 @@
     "emitDecoratorMetadata": true,
     "experimentalDecorators": true,
     "allowSyntheticDefaultImports": true,
-    "sourceMap": true,
+    "sourceMap": false,
+    "noEmit": false,
```
Comment (arranfw) [inline tsconfig.json:12]: Can we revert these changes if they weren't needed?

File: Makefile (hunk 6 above)
Comment (fwpushan) [reply Makefile]: DONE

File: tsconfig.json (hunk 7 above)
Comment (fwpushan) [reply tsconfig.json:12]: I'm keeping this for now to minimise build size.
